gluefactory/models/extractors/xfeat.py

```diff
@@ -0,0 +1,398 @@
+        )
+
+
+        ########### ⬇️ Fine Matcher MLP ⬇️ ###########
+        if self.model_name == ModelName.XFEAT_DENSE.value:
+            self.fine_matcher =  nn.Sequential(
+                nn.Linear(128, 512),
+                nn.BatchNorm1d(512, affine=False),
+                nn.ReLU(inplace = True),
+                nn.Linear(512, 512),
+                nn.BatchNorm1d(512, affine=False),
+                nn.ReLU(inplace = True),
+                nn.Linear(512, 512),
+                nn.BatchNorm1d(512, affine=False),
+                nn.ReLU(inplace = True),
+                nn.Linear(512, 512),
+                nn.BatchNorm1d(512, affine=False),
+                nn.ReLU(inplace = True),
+                nn.Linear(512, 64),
+            )
+
+    def _unfold2d(self, x: torch.Tensor, ws: int=2) -> torch.Tensor:
+        """
+            Unfolds tensor in 2D with desired ws (window size) and concat the channels
+        """
+        B, C, H, W = x.shape
+        x = x.unfold(2,  ws , ws).unfold(3, ws,ws)                             \
+            .reshape(B, C, H//ws, W//ws, ws**2)
+        return x.permute(0, 1, 4, 2, 3).reshape(B, -1, H//ws, W//ws)
+
+
+    def forward(self, x: torch.Tensor) -> dict:
+        """
+            input:
+                x -> torch.Tensor(B, C, H, W) grayscale or rgb images
+            return:
+                desc_map ->  torch.Tensor(B, 64, H/8, W/8) dense local features for descriptors
+                kpt_logit_map ->  torch.Tensor(B, 65, H/8, W/8) keypoint logit map
+                relmap   ->  torch.Tensor(B,  1, H/8, W/8) reliability map
+
+        """
+        #dont backprop through normalization
+        with torch.no_grad():
+            x = x.mean(dim=1, keepdim = True)
+            x = self.norm(x)
+
+        #main backbone
+        x1 = self.block1(x)
+        x2 = self.block2(x1 + self.skip1(x))
+        x3 = self.block3(x2)
+        x4 = self.block4(x3)
+        x5 = self.block5(x4)
+
+        #pyramid fusion
+        x4 = F.interpolate(x4, (x3.shape[-2], x3.shape[-1]), mode='bilinear')
+        x5 = F.interpolate(x5, (x3.shape[-2], x3.shape[-1]), mode='bilinear')
+        desc_map = self.block_fusion( x3 + x4 + x5 )
+
+        #heads
+        rel_map = self.heatmap_head(desc_map) # Reliability map
+        kpt_logit_map = self.keypoint_head(self._unfold2d(x, ws=8)) #Keypoint map logits
+
+        return {
+            'dense_descriptors':  desc_map,
+            'keypoint_logit_map': kpt_logit_map,
+            'reliability_map': rel_map
+        }
+
+
+class XFeat(BaseModel):
+    """ 
+        Implements the inference module for XFeat and XFeat-Dense models.
+    """
+    default_conf = {
+        "model_name": ModelName.XFEAT.value,
+        "max_num_keypoints": 4096,
+        "detection_threshold": 0.05,
+        "pretrained": True,
+        "get_sparse_outputs": True,
+        "force_num_keypoints": True,
+        "NMS": {
+            "threshold": 0.05,
+            "kernel_size": 5
```

Comment:
I think we are currently using nms of 3, should we do equal nms for all methods or keep them individual?
Also @hardik01shah make nms option explicit in demo config please and add a comment we are using the sparse xfeat.

Reply:
No, I think NMS should be method dependent.